Readme.md
# API for Dummy Data (RESTful)

Restful API fro dummy data built on node express


# Available resources
 - Posts
 - Todos
 - Users
 - Products

 # How to use 
 - [Post](https://github.com/bantawa04/redesigned-system-restful/wiki/Posts)
 - [Todo](https://github.com/bantawa04/redesigned-system-restful/wiki/Todo)
 - [Users](https://github.com/bantawa04/redesigned-system-restful/wiki/User)
 - [Products](https://github.com/bantawa04/redesigned-system-restful/wiki/Products)

CHANGELOG.md
#### 1.0.0 (2021-01-12)
- Readme.md added for documentation
- CHANGELOG.md added for tracking changes
##### Other Changes

* //github.com/bantawa04/redesigned-system-restful (fbcfb158)



test/Post.test.js
const chai = require('chai');
const chaiHttp = require('chai-http');
const app = require('../app');

const {expect} = chai;

chai.use(chaiHttp);

describe('Post API endpoint test', function () {
    it('Create new post',  function (done)  {
        chai
        .request(app)
        .post('/post')
        .send({title: "Title from mocha test", content: "content entered from mocha test"})
        .end(function(err, res){
            if(err) done(err);
            expect(res.status).to.equal(201);
            expect(res.body.post).to.be.an('object');
            console.log('Created Post');
            console.log(res.body.post);
            done();
        });
    })

    it('Delete a post', function(done) {
         chai.request(app)
        .delete('/post/5')
        .end(function(err, res){
            if(err) done(err);
            expect(res.status).to.equal(200);
            console.log('Post deleted');
            done();
        });
        
    });


    it('Get all 200 posts', function(done) {
        chai.request(app)
        .get('/post')
        .end(function(err, res){
            if(err) done(err);
            expect(res.status).to.equal(200);
            expect(res.body.posts).to.be.an('array');
            expect(res.body.posts).to.have.lengthOf(200);
            console.log('Length of posts array (200)');
            console.log(res.body.posts.length);
            done();
        })
    });

    it('Get 10 posts', function(done) {
        chai.request(app)
        .get('/post/?n=10')
        .end(function(err, res){
            if(err) done(err);
            expect(res.status).to.equal(200);
            expect(res.body.posts).to.be.an('array');
            expect(res.body.posts).to.have.lengthOf(10);
            done();
        })
    });

    it('Update post', function(done) {
        chai.request(app)
        .put('/post/18')
        .send({
            title: 'Updated post title',
            content: "Updated post content"
        })
        .end(function(err, res){
            if(err) done(err);
            expect(res.status).to.equal(200);
            expect(res.body.post).to.be.an('object');
            console.log('Updated post');
            console.log(res.body.post);
            done();
        })
    });    

    it('Get post by id 24', function(done) {
        chai.request(app)
        .get('/post/24')
        .end(function(err, res){
            if(err) done(err);
            expect(res.status).to.equal(200);
            expect(res.body.post).to.be.an('object');
            console.log('Post got by ID');
            console.log(res.body.post);
            done();
        })
    });

    it('No post above id 200', function(done) {
        chai.request(app)
        .get('/post/201')
        .end(function(err, res){
            if(err) done(err);
            expect(res.status).to.equal(422);
            console.log(res.body.message);
            done();
        })
    });

    it('Cannot update post above ID 200', function(done) {
        chai.request(app)
        .put('/post/201')
        .send({
            title: 'Updated post title',
            content: "Updated post content"
        })
        .end(function(err, res){
            if(err) done(err);
            expect(res.status).to.equal(400);
            done();
        })
    });    


});

test/Todo.test.js
const chai = require("chai");
const chaiHttp = require("chai-http");
const app = require("../app");

const { expect } = chai;

chai.use(chaiHttp);

describe("Post API endpoint test", function () {
  it("Create new todo", function (done) {
    chai
      .request(app)
      .post("/todo")
      .send({ title: "Title from mocha test", status: true })
      .end(function (err, res) {
        if (err) done(err);
        expect(res.status).to.equal(201);
        expect(res.body.todo).to.be.an("object");
        console.log(res.body.todo);
        done();
      });
  });


  it("Delete todo of id 33", function (done) {
    chai
      .request(app)
      .delete("/todo/33")
      .end(function (err, res) {
        if (err) done(err);
        expect(res.status).to.equal(200);
        done();
      });
  });

  it("Get all 200 todos", function (done) {
    chai
      .request(app)
      .get("/todo")
      .end(function (err, res) {
        if (err) done(err);
        expect(res.status).to.equal(200);
        expect(res.body.todos).to.be.an("array");
        expect(res.body.todos).to.have.lengthOf(200);        
        done();
      });
  });

  it("Get todo by ID 86", function (done) {
    chai
      .request(app)
      .get("/todo/86")
      .end(function (err, res) {
        if (err) done(err);
        expect(res.status).to.equal(200);
        expect(res.body.todo).to.be.an("object");
        console.log(res.body.todo);
        done();
      });
  });

    it('No todo above id 200', function(done) {
        chai.request(app)
        .get('/todo/201')
        .end(function(err, res){
            if(err) done(err);
            expect(res.status).to.equal(422);
            console.log(res.body);
            done();
        })
    });

    it('Cannot update post above ID 200', function(done) {
        chai.request(app)
        .put('/todo/201')
        .send({
            title: 'Todo title',
            status: false
        })
        .end(function(err, res){
            if(err) done(err);
            expect(res.status).to.equal(400);
            done();
        })
    });    

});


app.js
//Import Packages
const express = require("express");
const bodyParser = require("body-parser");
const cors = require("cors");
const helmet = require("helmet");
const compression = require("compression");

//Initialize express server
const app = express();

//Import routes
const postRoutes = require("./routes/Post");

const todoRoutes = require("./routes/Todo");

const userRoutes = require("./routes/User");

const productRoutes = require("./routes/Product");

//Middleware
app.use(cors());
app.use(helmet());
app.use(compression());
app.use(bodyParser.json());

//Implement Routes
app.use("/post", postRoutes);

app.use("/todo", todoRoutes);

app.use("/user", userRoutes);

app.use("/product", productRoutes);

//Catch error
app.use((error, req, res, next) => {
  console.log(error);
  const status = error.statusCode;
  const message = error.message;
  const data = error.data;
  res.status(status).json({ message: message, data: data });
});

//Run server at port 3000
app.listen(process.env.PORT || 3000);
module.exports = app;


routes/Product.js
const express = require("express");
const router = express.Router();
const { body } = require("express-validator");
const productController = require("../controller/ProductController");

router.get("/", productController.getProducts);
router.get("/:id", productController.getProductById);
router.post(
  "/",
  [
    body("title")
      .trim()
      .notEmpty()
      .isLength({ min: 5 })
      .withMessage("Title must be at least 5 chars long"),
    body("price")
      .notEmpty()
      .isNumeric(),
    body("description")
      .notEmpty()
      .escape()
      .isLength({ min: 5 })
      .withMessage("Description must be at least 5 chars long"),
  ],
  productController.createProduct
);

router.put(
  "/:id",
  [
    body("title")
      .trim()
      .notEmpty()
      .isLength({ min: 5 })
      .withMessage("Title must be at least 5 chars long"),
    body("price")
      .notEmpty()
      .isNumeric(),
    body("description")
      .notEmpty()
      .escape()
      .isLength({ min: 5 })
      .withMessage("Description must be at least 5 chars long"),
  ],
  productController.updateProduct
);

router.delete("/:id", productController.deleteProduct);
module.exports = router;


controller/TodoController.js
const { validationResult } = require("express-validator");
const faker = require("faker");
const { createTodo } = require("../utility/todo");

exports.getById = async (req, res, next) => {
  try {
    const todoId = req.params.id;
    if (todoId > 0 && todoId <= 200) {
      res.status(200).json({
        todo: { ...createTodo(), id: todoId },
      });
    } else {
      res.status(422).json({
        message: "Invalid request."
      });
    }
  } catch (err) {
    if (!err.statusCode) {
      err.statusCode = 500;
    }
    next(err);
  }
};

exports.getTodos = (req, res, next) => {
  let todos = [];
  try {
    if (req.query.n) {
      for (let i = 1; i <= req.query.n; i++) {
        todos.push({ ...createTodo(), id: i });
      }
    } else {
      for (let i = 1; i <= 200; i++) {
        todos.push({ ...createTodo(), id: i });
      }
    }
    res.status(200).json({
      todos: todos,
    });
  } catch (err) {
    if (!err.statusCode) {
      err.statusCode = 500;
    }
    next(err);
  }
};

exports.createTask = async (req, res, next) => {
  try {
    //   console.log(req);
    const errors = validationResult(req);
    const todo = {
      id: Math.floor(Math.random() * 201),
      task: req.body.title,
      status: req.body.status,
      created_at: faker.date.recent(),
    };
    if (!errors.isEmpty()) {
      const error = new Error(errors.array()[0].msg);
      error.statusCode = 422;
      throw error;
    }
    res.status(201).json({
      message: "Task created sucessfully !",
      todo: todo,
    });
  } catch (err) {
    if (!err.statusCode) {
      err.statusCode = 500;
    }
    next(err);
  }
};

exports.updateTodo = async (req, res, next) => {
  try {
    const id = req.params.id;
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      const error = new Error(errors.array()[0].msg);
      error.statusCode = 422;
      throw error;
    }

    if (id > 0 && id <= 200) {
      const todo = {
        id: id,
        title: req.body.title,
        status: req.body.status,
        created_at: faker.date.past(),
        updated_at: faker.date.recent(),
      };

      res.status(200).json({
        message: "Task updated",
        todo: todo,
      });
    } else {
      const error = new Error("Invalid input");
      error.message = "Todo with id "+id+" doesn't exist.";
      error.statusCode = 400;
      throw error;
    }
  } catch (err) {
    if (!err.statusCode) {
      err.statusCode = 500;
    }
    next(err);
  }
};

exports.deleteTodo = async (req, res, next) => {
  try {
    if (req.params.id > 0 && req.params.id <= 200) {
      res.status(200).json({
        message: "Task deleted successfully.",
      });
    } else {
      const error = new Error("Invalid input");
      error.message = "Task ID must be between 1 to 200";
      error.statusCode = 400;
      throw error;
    }
  } catch (err) {
    if (!err.statusCode) {
      err.statusCode = 500;
    }
    next(err);
  }
};


controller/UserController.js
const faker = require("faker");
const { createUser } = require("../utility/user");

exports.getUserById = async (req, res, next) => {
  try {
    const userId = req.params.id;
    if (userId > 0 && userId <= 200) {
      res.status(200).json({
        user: { ...createUser(), id: userId },
      });
    } else {
      const error = new Error("Invalid input");
      error.message = "Product with id "+userId+" doesn't exist.";
      error.statusCode = 400;
      throw error;;
    }
  } catch (err) {
    if (!err.statusCode) {
      err.statusCode = 500;
    }
    next(err);
  }
};

exports.getUsers = (req, res, next) => {
  let users = [];
  try {
    if (req.query.n) {
      for (let i = 1; i <= req.query.n; i++) {
        users.push({ ...createUser(), id: i });
      }
    } else {
      for (let i = 1; i <= 200; i++) {
        users.push({ ...createUser(), id: i });
      }
    }
    res.status(200).json({
      users: users,
    });
  } catch (err) {
    if (!err.statusCode) {
      err.statusCode = 500;
    }
    next(err);
  }
};


controller/PostController.js
const { validationResult } = require("express-validator");
const faker = require("faker");
const { createPost } = require("../utility/post");

exports.getPosts = (req, res, next) => {
  let posts = [];
  try {
    if (req.query.n) {
      for (let i = 1; i <= req.query.n; i++) {
        posts.push({ ...createPost(), id: i });
      }
    } else {
      for (let i = 1; i <= 200; i++) {
        posts.push({ ...createPost(), id: i });
      }
    }
    res.status(200).json({
      posts: posts,
    });
  } catch (err) {
    if (!err.statusCode) {
      err.statusCode = 500;
    }
    next(err);
  }
};

exports.createPost = async (req, res, next) => {
  try {
    //   console.log(req);
    const errors = validationResult(req);
    const post = {
      id: Math.floor(Math.random() * 201),
      title: req.body.title,
      content: req.body.content,
      created_at: faker.date.recent(),
    };
    if (!errors.isEmpty()) {
      const error = new Error(errors.array()[0].msg);
      error.statusCode = 422;
      throw error;
    }
    res.status(201).json({
      message: "Post created sucessfully !",
      post: post,
    });
  } catch (err) {
    if (!err.statusCode) {
      err.statusCode = 500;
    }
    next(err);
  }
};

exports.updatePost = async (req, res, next) => {
  try {
    const id = req.params.id;
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      const error = new Error(errors.array()[0].msg);
      error.statusCode = 422;
      throw error;
    }

    if (id > 0 && id <= 200) {
      const post = {
        id: id,
        title: req.body.title,
        content: req.body.content,
        created_at: faker.date.past(),
        updated_at: faker.date.recent(),
      };

      res.status(200).json({
        message: "Post updated",
        post: post,
      });
    } else {
      const error = new Error("Invalid input");
      error.message = "Post with id "+id+" doesn't exist.";
      error.statusCode = 400;
      throw error;
    }
  } catch (err) {
    if (!err.statusCode) {
      err.statusCode = 500;
    }
    next(err);
  }
};

exports.getPostById = async (req, res, next) => {
  try {
    const postId = req.params.id;

    if (postId < 200) {
      res.status(200).json({
        post: { ...createPost(), id: postId },
      });
    } else {
      res.status(422).json({
        message:"ID must be less than 200",
      });
    }
  } catch (err) {
    if (!err.statusCode) {
      err.statusCode = 500;
    }
    next(err);
  }
};

exports.deletePost = async (req, res, next) => {
  try {
    if (req.params.id > 0 && req.params.id <= 200) {
      res.status(200).json({
        message: "Post deleted successfully.",
      });
    } else {
      const error = new Error("Invalid input");
      error.message = "Post ID must be between 1 to 200";
      error.statusCode = 400;
      throw error;
    }
  } catch (err) {
    if (!err.statusCode) {
      err.statusCode = 500;
    }
    next(err);
  }
};
